62.CAD/main.cpp

```diff
@@ -276,49 +291,746 @@ class CPolyline
+		double p3 = p * p * p;
+		double q = kx * (2.0 * kx * kx - 3.0 * ky) + kz;
+		double h = q * q + 4.0 * p3;
+
+		if (h >= 0.0)
+		{
+			h = sqrt(h);
+			double2 x = (double2(h, -h) - q) / 2.0;
+
+			// Solving Catastrophic Cancellation when h and q are close (when p is near 0)
+			if (abs(abs(h / q) - 1.0) < 0.0001)
+			{
+				// Approximation of x where h and q are close with no carastrophic cancellation
+				// Derivation (for curious minds) -> h=√(q²+4p³)=q·√(1+4p³/q²)=q·√(1+w)
+				   // Now let's do linear taylor expansion of √(1+x) to approximate √(1+w)
+				   // Taylor expansion at 0 -> f(x)=f(0)+f'(0)*(x) = 1+½x 
+				   // So √(1+w) can be approximated by 1+½w
+				   // Simplifying later and the fact that w=4p³/q will result in the following.
+				x = double2(p3 / q, -q - p3 / q);
+			}
+
+			double2 uv = 
+				double2(x.X > 0.0 ? 1.0 : x.X == 0.0 ? 0.0 : -1.0, x.Y > 0.0 ? 1.0 : x.Y == 0.0 ? 0.0 : -1.0) // TODO get sign(x) working
+				* double2(pow(abs(x.X), 1.0 / 3.0), pow(abs(x.Y), 1.0 / 3.0));
+			double t = uv.X + uv.Y - kx;
+			t = std::clamp(t, 0.0, 1.0);
+
+			// 1 root
+			double2 qos = c + (2.0 * a + b * t) * t;
+			res = double2(sqrt(dot(qos, qos)), t);
+		}
+		else
+		{
+			double z = sqrt(-p);
+			double v = acos(q / (p * z * 2.0)) / 3.0;
+			double m = cos(v);
+			double n = sin(v) * 1.732050808;
+			double t[3] = { m + m, -n - m, n - m };
+			
+			for (uint32_t i = 0; i < 3; i++)
+				t[i] = std::clamp(t[i] * z - kx, 0.0, 1.0);
+
+			// 3 roots
+			double2 qos = c + (2.0 * a + b * t[0]) * t[0];
+			double dis = dot(qos, qos);
+
+			res = double2(dis, t[0]);
+
+			qos = c + (2.0 * a + b * t[1]) * t[1];
+			dis = dot(qos, qos);
+			if (dis < res.X) res = double2(dis, t[0]);
+
+			qos = c + (2.0 * a + b * t[2]) * t[2];
+			dis = dot(qos, qos);
+			if (dis < res.X) res = double2(dis, t[2]);
+
+			res.X = sqrt(res.X);
+		}
+
+		return res.X;
+	}
+
+	class Segment 
+	{
+	public:
+		const QuadraticBezierInfo* originalBezier;
+		// because beziers are broken down,  depending on the type this is t_start or t_end
+		double t_start;
+		double t_end; // beziers get broken down
+
+		QuadraticBezierInfo getSplitCurve()
+		{
+			double2 a = originalBezier->p[0] - 2.0 * originalBezier->p[1] + originalBezier->p[2];
+			double2 b = 2.0 * (originalBezier->p[1] - originalBezier->p[0]);
+			double2 c = originalBezier->p[0];
+
+			return { 
+				a * (t_end - t_start) * (t_end - t_start),
+				(t_end - t_start) * (2 * a * t_start + b),
+				Hatch::evaluteBezier(originalBezier, t_start)
+			};
+		}
+
+		double intersect(const Segment& other) const
+		{
+			const std::array<double, 2> intersections = Hatch::intersectionAlgorithm(originalBezier, other.originalBezier);
+			for (uint32_t i = 0; i < 2; i++)
+			{
+				auto t = intersections[i];
+				if (other.t_start >= t || t >= other.t_end)
+					continue;
+
+				auto evaluated = Hatch::evaluteBezier(other.originalBezier, t);
+				auto thisT = Hatch::closestTInCurve(originalBezier, evaluated);
```

Comment:
Holy fuck, I know why, you find the "t" value for the other bezier. and then evaluate the other bezier at that point and then do an SDF to get the original bezier t, 
1. since you're returning the t value for the original bezier it's best to compute that first and then check if that point is also on the other "implicitized curve"
2. Using only the t for sdf is wrong as it returns the t value corresponding to the closest point on the bezier. and that t closest point might be miles away :D but in the range of [0,1]

Reply:
So
1. get the t in the intersection for the original bezier and if it is in [0,1] ->
2. then check if that point is within some threshold on the other bezier as well. (details below in how to do this)
3. If it is then return the found t

so how do you figure out if the found intersection point is also on the other bezier as well? 
generally you would use the SDF like here and use BOTH the unclamped t value and the DISTANCE to make sure it's not outside the bezier, but we have some many requirements of our beziers when they get here. they are:
1. monotonically increasing in the direction of the major axis
2. you already found the intersection point, it's just that it should be within it's start and end, it's not some random point.

Reply:
messaged on discord about how you can do this MUCH more simpler